### PR TITLE
Class Cog Overhaul

### LINKED ISSUE #3: Translate Cog: Missing key
It seems SockBot is missing the key for the translate command. 
image
@ me or DM me and we can get the code added to it (if I can remember where in Azure it is)

### PR DESCRIPTION
**Pull request by jkriste**
## Overlook
- Class channels will be REUSED.
  - When a semester starts, class creation will unarchive a previous class, if it exists.
  - This allows students to retain:
    - Class pins.
    - Thoughts/notes/advice from previous students.
- Semesters will be hard-coded with start and end dates. 
  - The end date will facilitate automatic archival of classes.
  - The start date will facilitate students' ability to create/unarchive classes.
  - A semester can be archived manually, if necessary.
- Students that have the class role can add a "pin request" that, given X amount of votes/reactions will be pinned in the class channel.
  - Students can either reply to the message they want pinned, or supply the link to the message.
  - An embed with the content, link, original message author, user that requested the pin will be sent with a 📌 reaction.
  - If the total pins in the class reaches 50 (the maximum amount of pins that can belong to a specific text channel) and a pin request was fulfilled, SockBot will unpin the oldest pin in exchange for the new pin.
- On creation or unarchival of a semester, SockBot will send an embed to notify the creator/requester the channel was created.
  - Title: 📔 **PREF-####: Class Name**
  - Message: Welcome, students! Click the ✅ reaction below to join the class.
  - Fields: Semester, Professor, Creator
  - On a user's ✅ reaction, SockBot will apply the class role to the user (if not already applied).
- SockBot will listen to channel deletion. If the deleted channel was a class, it will remove it from the database.
- SockBot will listen to message deletion. If a reaction post or class pin was deleted, it will remove it from the database.
- Classes that are created will be reformatted to `prefix-number-lastname`. Spaces will not be reformatted to `-` and only the last name of the professor will be used, instead of the name the user gives us (ex: `cpsc-4910-van-scoy`)

### Commands
`?class add` - Already implemented, but pending improvements.
`?class semester` - Lets the user know if there is a current semester and provides the start and end dates. If not, provides the user with the start date of the upcoming semester.
`?class semester archive [semester_id]` - Manual archival of a current or previous semester. Requires a claim.
`?class archive <channel_id_or_mention>` - Manual archival of a class channel. Requires a claim.
`?class insert <channel_id_or_mention>` - Manual insertion of a class. Will fail if the class already exists in the database. Requires a claim.
`?class check <channel_id_or_mention>` - Checks if the class is a part of the database. Requires a claim.

Fixes #3

### CODE REVIEW

**Comment by jkriste** [comment]
Things not mentioned in the description that have changed since it was written:
- All `?class` commands mentioned have been changed to slash-commands.
- Individual class channels that have been archived can be un-archived, but entire semesters cannot (is there even a use-case for this?).

Tips for testing:
- "ClassArchiveCategoryIds" and "ClassNotifsChannelId" in the `BotSecrets.json` file will have to be added AND set before running the bot. Otherwise, things will fail catastrophically.
- For the `?pin` command, when clicking the 📌 reaction on a pin request, if you are either part of the Clemson CPSC staff or have administrator permissions, the pin request will be approved before the amount of reactions (5 total reactions = pin) reaches 5. This is by design.

File: bot/errors.py [hunk 1]
```diff
@@ -42,3 +42,12 @@ class ConversionError(CommandError):
 
     def __init__(self, message):
         self.message = message
+
+
+class NoSemesterError(CommandError):
```
Comment (jkriste) [inline bot/errors.py:47]: todo: remove this.

File: bot/messaging/events.py [hunk 2]
```diff
@@ -475,5 +475,74 @@ def on_member_ban(self):
...
+        return 'on_pin_request'
+
+    @property
+    def on_class_unarchive(self):
+        """
+        Published when a user requests to create a previously-created class.
+
+        Args:
+            inter (discord.Interaction): The app command interaction.
+            pref (str): Class prefix, i.e., CPSC, MATH, HCC
```
Comment (jkriste) [inline bot/messaging/events.py:528]: todo: update this documentation

**Review by Iapetus-11 — APPROVED** [review]
lgtm

File: bot/data/base_repository.py [hunk 3]
```diff
@@ -46,7 +45,7 @@ async def fetcthall_as_class(self, cursor: aiosqlite.Cursor):
         return [type('Query', (), dict(zip([column[0] for column in cursor.description], row)))
```
Comment (Iapetus-11) [inline bot/data/base_repository.py:45]: for FUCKS sake just use an ORM

**Review by Iapetus-11 — CHANGES_REQUESTED** [review]

File: bot/cogs/manage_classes_cog.py [hunk 4]
```diff
@@ -1,365 +1,172 @@
...
-                value="cpsc-1010",
-                inline=False,
-            )
+    @app_commands.command(name='insert', description='Insert a class channel.')
+    async def insert(self, inter: discord.Interaction, channel: discord.TextChannel, role: discord.Role | None = None):
+        # check perms
+        if not await self._perms_check(inter):
+            return
+        # check if a current semester exists
+        if not await self.repo.get_current_semester():
```
Comment (Iapetus-11) [inline bot/cogs/manage_classes_cog.py:53]: Yknow code is more readable and looks nicer when spaced out reasonably. General rule I follow is a new line after dedentation occurs.

File: bot/cogs/pin_cog.py [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+        if not message and ctx.message.reference:
+            message = await ctx.channel.fetch_message(ctx.message.reference.message_id)
+        if not message:
+            await deletable_error_embed(self.bot, ctx, 'Reply to a message or give the ID or link to a message.')
+            return
+        if message.pinned:
+            await deletable_error_embed(self.bot, ctx, 'The message given is already pinned.')
+            return
+        if message.author.bot:
+            await deletable_error_embed(self.bot, ctx, 'Cannot pin a bot message.')
```
Comment (Iapetus-11) [inline bot/cogs/pin_cog.py:37]: Why not

File: bot/data/class_repository.py [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+
+    async def get_next_semester(self) -> ClassSemester | None:
+        """
+        Gets the next semester, if one exists.
+        """
+        current_datetime = datetime.datetime.utcnow()
+        for semester in await self.get_all_semesters():
+            if current_datetime < semester.start_date():
+                return semester
+        return None
```
Comment (Iapetus-11) [inline bot/data/class_repository.py:40]: This could be done in sql

File: bot/data/class_repository.py [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+        Searches for a registered class given the class abbreviation, class number, and class instructor.
+        """
+        async with aiosqlite.connect(self.resolved_db_path) as db:
+            cursor = await db.execute(
+                """ SELECT * FROM ClassChannel 
+                    WHERE class_prefix = ? 
+                    AND class_number = ? 
+                    AND class_professor = ?""", (prefix, num, prof)
+            )
+            dictionary = await self.fetcthone_as_dict(cursor)
```
Comment (Iapetus-11) [inline bot/data/class_repository.py:85]: fetcthone

File: bot/modals/class_modal.py [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+        row=4
+    )
+
+    def __init__(self, bot: SockBot,
+                 *,
+                 class_data: tuple[str, int] | None = None,
+                 channel: discord.TextChannel | None = None,
+                 role: discord.Role | None = None):
+        super().__init__(title='📔 Add Class' if not channel else f'Insert #{channel.name}')
+        self._bot = bot
```
Comment (Iapetus-11) [inline bot/modals/class_modal.py:76]: Is a `bot` attribute not already set by the superclass?

File: bot/cogs/manage_classes_cog.py (hunk 4 above)
Comment (jkriste) [reply bot/cogs/manage_classes_cog.py:53]: Y'know I've always been the kind of person to keep all of my code compact for some reason, and I can't seem to shake the habit. I was hoping the comments would be a kind of "compromise".

File: bot/cogs/pin_cog.py (hunk 5 above)
Comment (jkriste) [reply bot/cogs/pin_cog.py:37]: I do not see a situation in a class channel that a student would have to pin a message from a bot, but if you can think of one I will reconsider it. :)

File: bot/data/class_repository.py (hunk 7 above)
Comment (jkriste) [reply bot/data/class_repository.py:85]: When I was first starting this PR I was really confused as to why it was named this as well. I'll probably go back and make a better name for them...

File: bot/modals/class_modal.py (hunk 8 above)
Comment (jkriste) [reply bot/modals/class_modal.py:76]: Nah Modals do not have a `bot` attribute, so in order to call the events I needed to pass the bot instance to it.

File: bot/cogs/pin_cog.py [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+
+class PinCog(commands.Cog):
+
+    def __init__(self, bot: SockBot) -> None:
+        self.bot = bot
+        self.class_repo = ClassRepository()
+        self.pin_repo = PinRepository()
+
+    @ext.command(case_insensitive=True)
+    # @commands.cooldown(1, PIN_COMMAND_COOLDOWN, commands.BucketType.user)
```
Comment (jkriste) [inline bot/cogs/pin_cog.py:23]: todo: remove line + `PIN_COMMAND_COOLDOWN`

File: bot/cogs/manage_classes_cog.py [hunk 10]
```diff
@@ -1,365 +1,181 @@
...
+            embed = error_embed(inter.user, f'The given channel {channel.mention} is already registered.')
+            await inter.response.send_message(embed=embed, ephemeral=True)
+            return
+        await inter.response.send_modal(AddClassModal(self.bot, channel=channel, role=role))
 
-            await ctx.send(embed=embed)
+    @app_commands.command(name='archive', description='Manually archive a class channel.')
+    async def class_archive(self, inter: discord.Interaction, channel: discord.TextChannel):
+        # check perms
+        if not await self._perms_check(inter):
```
Comment (Iapetus-11) [inline bot/cogs/manage_classes_cog.py:72]: Can't we use d.py checks https://discordpy.readthedocs.io/en/stable/ext/commands/commands.html#checks I think they work with slash commands too

File: bot/utils/helpers.py [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+    return datetime.strptime(date, FORMAT)
+
+
+def as_timestamp(date: datetime, /, style: Literal["f", "F", "d", "D", "t", "T", "R"] = "f") -> str:
+    """
+    Formats the given datetime to a Discord timestamp.
+    Used over discord.utils.format_dt due to incorrect timestamp output.
+    """
+    timestamp = calendar.timegm(date.utctimetuple())
+    return f'<t:{timestamp}:{style}>'
```
Comment (Iapetus-11) [inline bot/utils/helpers.py:26]: This is already part of discord.py https://discordpy.readthedocs.io/en/stable/api.html#discord.utils.format_dt

File: bot/utils/helpers.py [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+from typing import Literal, Union
+
+import discord
+from discord import NotFound
+
+from bot.consts import Colors
+from bot.messaging.events import Events
+from bot.sock_bot import SockBot
+
+FORMAT = '%Y-%m-%d %H:%M:%S'
```
Comment (Iapetus-11) [inline bot/utils/helpers.py:12]: This should be more descriptive, like `DATE_FORMAT`

File: bot/services/pin_service.py [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+                await react.message.delete()
+                await self.pin_repo.delete_pin(class_pin)
+                return
+            # check if we have enough space to pin
+            pinned_messages = await channel.pins()
+            # unpin the oldest pin if there is no room
+            if len(pinned_messages) == MAX_PINS_PER_CHANNEL:
+                await pinned_messages[0].unpin(reason='Unpinned to make room.')
+            # pin the message, update the pin in the db, and delete sockbot's pin request embed
+            await to_pin.pin(reason='Pinned by `pin` command.')
```
Comment (Iapetus-11) [inline bot/services/pin_service.py:72]: What about specifying the person who started the vote? Also I'm not sure if markdown works in the audit log

File: bot/services/pin_service.py [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+
+PIN_REACTION = '📌'
+MIN_PIN_REACTIONS = 5
+MAX_CONTENT_CHARS = 250
+MAX_PINS_PER_CHANNEL = 50
+
+
+class PinService(BaseService):
+
+    def __init__(self, *, bot: SockBot):
```
Comment (Iapetus-11) [inline bot/services/pin_service.py:24]: What's the point of the * here

File: bot/services/class_service.py [hunk 15]
```diff
@@ -0,0 +1,392 @@
...
+from bot.data.class_repository import ClassRepository
+from bot.messaging.events import Events
+from bot.models.class_models import ClassSemester, ClassChannel, ClassChannelScaffold
+from bot.services.base_service import BaseService
+from bot.sock_bot import SockBot
+from bot.utils.helpers import error_embed, fetch_optional_message
+
+log = logging.getLogger(__name__)
+MAX_CHANNELS_PER_CATEGORY = 50
+WELCOME_MESSAGE_REACTION = '✅'
```
Comment (Iapetus-11) [inline bot/services/class_service.py:17]: IMO I would put log below these constants and separate them with an extra newline but that's IMO

File: bot/models/class_models.py [hunk 16]
```diff
@@ -0,0 +1,59 @@
...
+
+    def full_title(self) -> str:
+        return f'{self.class_code()}: {self.class_name}'
+
+    def channel_name(self) -> str:
+        return f'{self.class_code().lower()}-{self.class_professor.lower()}'
+
+    def intended_category(self) -> str:
+        level = (self.class_number // 1000) * 1000
+        return f'{self.class_prefix} {level} LEVELS'
```
Comment (Iapetus-11) [inline bot/models/class_models.py:39]: Don't do this, use f-string formatting 
image

File: bot/models/class_models.py [hunk 17]
```diff
@@ -0,0 +1,59 @@
...
+    class_name: str
+
+    def class_code(self) -> str:
+        return f'{self.class_prefix}-{self.class_number}'
+
+    def full_title(self) -> str:
+        return f'{self.class_code()}: {self.class_name}'
+
+    def channel_name(self) -> str:
+        return f'{self.class_code().lower()}-{self.class_professor.lower()}'
```
Comment (Iapetus-11) [inline bot/models/class_models.py:35]: Why not make these properties?

File: bot/modals/class_modal.py [hunk 18]
```diff
@@ -0,0 +1,164 @@
...
+        if len(split_name) >= 3:
+            if valid_course_maj(split_name[0]):
+                self.major.default = split_name[0].upper()
+            if valid_course_num(split_name[1]):
+                self.course_number.default = split_name[1].strip()
+            self.professor.default = split_name[-1].strip().capitalize()[:MAX_INSTR_LEN]
+        if len(split_topic) == 2:
+            self.course_title.default = split_topic[0].title().strip()[:MAX_TITLE_LEN]
+            description = split_topic[1].strip()[:MAX_DESCR_LEN]
+            self.course_description.default = description if len(description) != 0 else 'None'
```
Comment (Iapetus-11) [inline bot/modals/class_modal.py:149]:
```suggestion
            self.course_description.default = description or 'None'
```

File: bot/modals/class_modal.py [hunk 19]
```diff
@@ -0,0 +1,164 @@
...
+    async def _search_similar(self, inter: discord.Interaction, pref: str, num: int, prof: str) -> bool:
+        similar_class = await self._repo.search_class(pref, num, prof)
+        if not similar_class:
+            return False
+        channel = self._bot.guild.get_channel(similar_class.channel_id)
+        if channel and not similar_class.class_archived:
+            if role := self._bot.guild.get_role(similar_class.class_role_id):
+                await inter.user.add_roles(role)
+            embed = discord.Embed(title='📔 Similar Class Found', color=Colors.Purple)
+            embed.description = f'A similar class channel has been found.'
```
Comment (Iapetus-11) [inline bot/modals/class_modal.py:119]: unnecessary f string

File: bot/modals/class_modal.py [hunk 20]
```diff
@@ -0,0 +1,164 @@
...
+            self._autofill(channel if channel else class_data)
+
+    async def on_submit(self, inter: Interaction) -> None:
+        # correct for error - reject if invalid
+        if not valid_course_num(self.course_number.value):
+            embed = error_embed(inter.user, f'Course number `{self.course_number.value}` is invalid.')
+            await inter.response.send_message(embed=embed, ephemeral=True)
+            return
+        # format our data, so it's nice and neat and capitalized correctly (I know you guys don't use caps)
+        professor = self.professor.value.split(' ')[-1].capitalize()
```
Comment (Iapetus-11) [inline bot/modals/class_modal.py:90]: split defaults to a string with one space, remove the unnecessary param

File: bot/messaging/events.py [hunk 21]
```diff
@@ -475,5 +475,76 @@ def on_member_ban(self):
...
+    @property
+    def on_semester_archive(self):
+        """
+        Published when a semester is manually archived or scheduled to be archived.
+
+        Args:
+            semester (bot.models.class_models.ClassSemester): The semester to archive.
+            inter (discord.Interaction | None = None): The app command interaction.
+        """
+        return 'on_semester_archive'
```
Comment (Iapetus-11) [inline bot/messaging/events.py:487]: I fucking hate this events class I wish there was a better way of doing this

File: bot/cogs/manage_classes_cog.py [hunk 22]
```diff
@@ -1,365 +1,201 @@
...
-
-
-def round_down(num: int, divisor: int) -> int:
-    return num - (num % divisor)
+            embed = error_embed(inter.user, 'Could not find current nor next semester.')
+            await inter.response.send_message(embed=embed)
+
+    @semester_group.command(name='archive', description='Manually archive a semester.')
+    async def semester_archive(self, inter: discord.Interaction, semester_id: str):
+        # check perms
```
Comment (Iapetus-11) [inline bot/cogs/manage_classes_cog.py:175]: This comment is dumb

File: bot/cogs/manage_classes_cog.py (hunk 10 above)
Comment (jkriste) [reply bot/cogs/manage_classes_cog.py:77]: Yes, I tested this out, but instead of something simple, like the bot telling the user they don't have permission, it will just raise an uncaught exception.

File: bot/data/CreateTables.sql [hunk 23]
```diff
@@ -0,0 +1,65 @@
...
+-- Populate ClassSemester table, if the values do not exist.
+
+-- !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+-- The start and end dates have been adjusted by approximately *5 DAYS* of additional   !!
+-- time before and after the start and end dates for preparation/final discussion.      !!
+-- Times formatted as UTC (5 AM UTC is 12 AM EST), ignore daylight savings time changes.!!
+-- Start AND end date times should NOT overlap with other semesters.                    !!
+-- !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+
+INSERT OR IGNORE INTO ClassSemester(semester_id, semester_name, semester_start, semester_end) VALUES
```
Comment (Iapetus-11) [inline bot/data/CreateTables.sql:62]: this is cool

File: bot/utils/helpers.py (hunk 11 above)
Comment (jkriste) [reply bot/utils/helpers.py:26]: Their method actually ends up using GMT instead of UTC as the timezone, which heavily skews the actual timestamp Discord will display, so this is why this method exists.

File: bot/services/pin_service.py (hunk 13 above)
Comment (jkriste) [reply bot/services/pin_service.py:75]: Good idea, also figured out markdown does not work in the audit log lmao

File: bot/modals/class_modal.py [hunk 24]
```diff
@@ -0,0 +1,196 @@
...
+
+def valid_course_num(course_num: Union[str, int]) -> bool:
+    number: int
+    if isinstance(course_num, str):
+        if not course_num.isdigit():
+            return False
+        number = int(course_num)
+    else:
+        number = course_num
+    return MIN_CLASS_NUM <= number <= MAX_CLASS_NUM
```
Comment (Iapetus-11) [inline bot/modals/class_modal.py:196]: int int int
```suggestion
    number: int
    if isinstance(course_num, str) and not course_num.isdigit():
        return False
        
    return MIN_CLASS_NUM <= int(course_num) <= MAX_CLASS_NUM
```

File: bot/data/class_repository.py (hunk 7 above)
Comment (Iapetus-11) [reply bot/data/class_repository.py]: fetchone ?

File: bot/cogs/manage_classes_cog.py (hunk 10 above)
Comment (Iapetus-11) [reply bot/cogs/manage_classes_cog.py:77]: Yes which is why you catch it :suspect:

File: bot/utils/helpers.py (hunk 11 above)
Comment (Iapetus-11) [reply bot/utils/helpers.py:26]: image

File: bot/services/pin_service.py (hunk 14 above)
Comment (jkriste) [reply bot/services/pin_service.py:24]: Hm, good question. It was like this in the `example_service.py` so I just followed it. I'll remove it.

**Comment by Iapetus-11** [comment]
mmm stupid sexy flanders

File: bot/models/class_models.py (hunk 16 above)
Comment (jkriste) [reply bot/models/class_models.py:39]: This has the downside of making it basically unreadable, plus it's taking an already large number (ex. `2120`, `4910`) and getting the thousandths place and adding three 0's basically. Not sure how you would do that with f-string formatting.

File: bot/models/class_models.py (hunk 17 above)
Comment (jkriste) [reply bot/models/class_models.py:35]: Because they are read-only and are just for formatting.

File: bot/modals/class_modal.py (hunk 18 above)
Comment (jkriste) [reply bot/modals/class_modal.py:181]: Oooh, you learn something new every day. Thanks.

File: bot/modals/class_modal.py (hunk 19 above)
Comment (jkriste) [reply bot/modals/class_modal.py:130]: Great catch

File: bot/models/class_models.py (hunk 16 above)
Comment (Iapetus-11) [reply bot/models/class_models.py:39]: ah shiii i misread 1 sec

File: bot/models/class_models.py (hunk 17 above)
Comment (Iapetus-11) [reply bot/models/class_models.py:35]: Wroooong

```py
@property
def bruh():
   return "This shit unsettable"
```

File: bot/modals/class_modal.py (hunk 20 above)
Comment (jkriste) [reply bot/modals/class_modal.py:90]: Neat, didn't know that

File: bot/messaging/events.py (hunk 21 above)
Comment (jkriste) [reply bot/messaging/events.py:487]: LMAO if you can think of any better way I'll def. consider making a PR for it

File: bot/cogs/manage_classes_cog.py (hunk 22 above)
Comment (jkriste) [reply bot/cogs/manage_classes_cog.py:175]: based, chadly, removed 🇹🇩

File: bot/modals/class_modal.py (hunk 24 above)
Comment (jkriste) [reply bot/modals/class_modal.py:196]: int this, int that, intellectually you are bad (changed, thanks)

File: bot/data/class_repository.py (hunk 7 above)
Comment (jkriste) [reply bot/data/class_repository.py]: more like fetch_first 😈

File: bot/utils/helpers.py (hunk 11 above)
Comment (jkriste) [reply bot/utils/helpers.py:26]: fake news. i'll go ahead and test..........aha image
```python
embed.add_field(name='Start Date (og)', value=as_timestamp(current_semester.start_date))
embed.add_field(name='End Date (og)', value=as_timestamp(current_semester.end_date))
embed.add_field(name='Start Date (discord)', value=format_dt(current_semester.start_date))
embed.add_field(name='End Date (discord)', value=format_dt(current_semester.end_date))
```

**Review by Iapetus-11 — APPROVED** [review]
I've done all the reviewing I can possibly muster

TLDR: LGTM

**Review by Iapetus-11 — APPROVED** [review]
sure

**Comment by Jay-Madden** [comment]
Jesus christ yall lol, good reviews

File: bot/modals/class_modal.py [hunk 25]
```diff
@@ -0,0 +1,192 @@
...
+MIN_INSTR_LEN = 2
+MAX_INSTR_LEN = 15
+# Max length of class description, i.e., 'Students learn about...'
+MAX_DESCR_LEN = 500
+# Min and max of class number, i.e., 1060, 4910, 8400, etc.
+MIN_CLASS_NUM = 1000
+MAX_CLASS_NUM = 8999
+
+
+class AddClassModal(Modal):
```
Comment (Jay-Madden) [inline bot/modals/class_modal.py:30]: This needs to implement on_error() so that we can log the errors in the error channel

Comment (Jay-Madden) [reply bot/modals/class_modal.py:30]: probably need to pass the error callback through the modal object creation in manage_classes_cog

Comment (Jay-Madden) [reply bot/modals/class_modal.py:30]: image

File: bot/services/class_service.py [hunk 26]
```diff
@@ -0,0 +1,433 @@
...
+        """
+        Fetches the 'Cleanup' role from the guild.
+        If not found, a new role with the name 'Cleanup' is created and returned.
+        """
+        for role in self.bot.guild.roles:
+            if role.name == 'Cleanup':
+                return role
+        return await self.bot.guild.create_role(name='Cleanup')
+
+    async def _send_welcome(self, cls: ClassChannel, just_created: bool, user: discord.User):
```
Comment (Jay-Madden) [inline bot/services/class_service.py:340]: The welcome should also ping the user that created it so that they can find the channel

**Review by Jay-Madden — CHANGES_REQUESTED** [review]

File: bot/cogs/manage_classes_cog.py [hunk 27]
```diff
@@ -1,365 +1,231 @@
...
-    async def archive(self, ctx, channel: discord.TextChannel) -> None:
-        pass
-
-    async def input_timeout(self, ctx) -> None:
-        await ctx.send("Response timed out please redo the class wizard")
-
-
-def round_down(num: int, divisor: int) -> int:
-    return num - (num % divisor)
+    async def _on_modal_error(self, inter: discord.Interaction, error: Exception) -> None:
```
Comment (Jay-Madden) [inline bot/cogs/manage_classes_cog.py:224]: put this in sock_bot.py

**Review by Jay-Madden — APPROVED** [review]
Lets see if this works
